fraud/service.go

```diff
@@ -23,20 +25,39 @@ type service struct {
 	topicsLk sync.RWMutex
 	topics   map[ProofType]*topic
 
+	storeLk sync.RWMutex
+	stores  map[ProofType]datastore.Datastore
+
 	pubsub *pubsub.PubSub
 	getter headerFetcher
+	ds     datastore.Datastore
 }
 
-func NewService(p *pubsub.PubSub, getter headerFetcher) Service {
+func NewService(p *pubsub.PubSub, getter headerFetcher, ds datastore.Datastore) Service {
 	return &service{
 		pubsub: p,
 		getter: getter,
 		topics: make(map[ProofType]*topic),
+		stores: make(map[ProofType]datastore.Datastore),
+		ds:     ds,
 	}
 }
 
 func (f *service) Subscribe(proofType ProofType) (Subscription, error) {
-	// TODO: @vgonkivs check if fraud proof is in fraud store, then return with error
+	f.storeLk.Lock()
+	if _, ok := f.stores[proofType]; !ok {
+		f.stores[proofType] = namespace.Wrap(f.ds, makeKey(proofType))
+	}
+	store := f.stores[proofType]
+	proofs, err := getAll(context.TODO(), store)
+	f.storeLk.Unlock()
+	if err == nil {
+		return nil, &ErrFraudExists{Type: proofType, Proofs: proofs}
```

Comment:
What is supposed to happen with this particular error type? I do not see it used anywhere besides in the test.

Reply:
Yes

Reply:
I commented over the same thing pointing to where it should be handled